{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"name":"RawSummarizeBART.ipynb","provenance":[],"collapsed_sections":[],"toc_visible":true,"mount_file_id":"1jsm2lT4Aa7--XSB0_Rd1MpZYtDrO7-I7","authorship_tag":"ABX9TyPIZoQkqghjUG1revMXWH4R"},"kernelspec":{"name":"python3","display_name":"Python 3"},"accelerator":"GPU"},"cells":[{"cell_type":"code","metadata":{"id":"iqjCDAC5UUvN","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":17},"outputId":"78403e55-a585-4d20-e68a-882adef7d697","executionInfo":{"status":"ok","timestamp":1589121087176,"user_tz":-120,"elapsed":4510,"user":{"displayName":"Bartłomiej Cerek","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GgghwbmOdc_vcrQTrbQaIvP0ASodbJHQUXxGgJyyF4=s64","userId":"03865172822680659937"}}},"source":["%%capture\n","!pip install transformers"],"execution_count":38,"outputs":[{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}}]},{"cell_type":"code","metadata":{"id":"wbGC9_gLYT8o","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":17},"outputId":"c6bab106-ad03-4eb9-8aca-779cf9887eef","executionInfo":{"status":"ok","timestamp":1589121087178,"user_tz":-120,"elapsed":4487,"user":{"displayName":"Bartłomiej Cerek","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GgghwbmOdc_vcrQTrbQaIvP0ASodbJHQUXxGgJyyF4=s64","userId":"03865172822680659937"}}},"source":["# Wrap cels\n","from IPython.display import HTML, display\n","\n","def set_css():\n","  display(HTML('''\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  '''))\n","get_ipython().events.register('pre_run_cell', set_css)"],"execution_count":39,"outputs":[{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}}]},{"cell_type":"code","metadata":{"id":"Xd209r0ZUm7d","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":17},"outputId":"51826fbe-a243-49ce-ceb2-b9837ff7fb7c","executionInfo":{"status":"ok","timestamp":1589121087178,"user_tz":-120,"elapsed":4478,"user":{"displayName":"Bartłomiej Cerek","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GgghwbmOdc_vcrQTrbQaIvP0ASodbJHQUXxGgJyyF4=s64","userId":"03865172822680659937"}}},"source":["import numpy as np\n","import pandas as pd\n","\n","from transformers import BartTokenizer, BartForConditionalGeneration, BartConfig\n","\n","import torch"],"execution_count":40,"outputs":[{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}}]},{"cell_type":"code","metadata":{"id":"Pon0SMNpUvoT","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":459},"outputId":"461e412e-eca7-45ad-edb2-80a3ff3a52dd","executionInfo":{"status":"ok","timestamp":1589121088223,"user_tz":-120,"elapsed":5514,"user":{"displayName":"Bartłomiej Cerek","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GgghwbmOdc_vcrQTrbQaIvP0ASodbJHQUXxGgJyyF4=s64","userId":"03865172822680659937"}}},"source":["articles_df = pd.read_csv('drive/My Drive/!UBS/DW/data/covid_main.csv', index_col=0, dtype ={'body': str})\n","print(articles_df.shape)\n","articles_df.head()"],"execution_count":41,"outputs":[{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"stream","text":["(1933, 8)\n"],"name":"stdout"},{"output_type":"execute_result","data":{"text/html":["<div>\n","<style scoped>\n","    .dataframe tbody tr th:only-of-type {\n","        vertical-align: middle;\n","    }\n","\n","    .dataframe tbody tr th {\n","        vertical-align: top;\n","    }\n","\n","    .dataframe thead th {\n","        text-align: right;\n","    }\n","</style>\n","<table border=\"1\" class=\"dataframe\">\n","  <thead>\n","    <tr style=\"text-align: right;\">\n","      <th></th>\n","      <th>sha</th>\n","      <th>title</th>\n","      <th>abstract</th>\n","      <th>publish_time</th>\n","      <th>authors</th>\n","      <th>url</th>\n","      <th>introduction</th>\n","      <th>body</th>\n","    </tr>\n","  </thead>\n","  <tbody>\n","    <tr>\n","      <th>31175</th>\n","      <td>f056da9c64fbf00a4645ae326e8a4339d015d155</td>\n","      <td>SIANN: Strain Identification by Alignment to N...</td>\n","      <td>Next-generation sequencing is increasingly bei...</td>\n","      <td>2014-01-10</td>\n","      <td>Samuel Minot; Stephen D Turner; Krista L Ternu...</td>\n","      <td>https://doi.org/10.1101/001727</td>\n","      <td>There are many different methods that characte...</td>\n","      <td>Metagenomic classification methods are based o...</td>\n","    </tr>\n","    <tr>\n","      <th>31176</th>\n","      <td>daf32e013d325a6feb80e83d15aabc64a48fae33</td>\n","      <td>Spatial epidemiology of networked metapopulati...</td>\n","      <td>An emerging disease is one infectious epidemic...</td>\n","      <td>2014-06-04</td>\n","      <td>Lin WANG; Xiang Li</td>\n","      <td>https://doi.org/10.1101/003889</td>\n","      <td>The term metapopulation was coined by Levins [...</td>\n","      <td>Due to the ubiquity of complex systems in mode...</td>\n","    </tr>\n","    <tr>\n","      <th>31177</th>\n","      <td>f33c6d94b0efaa198f8f3f20e644625fa3fe10d2</td>\n","      <td>Sequencing of the human IG light chain loci fr...</td>\n","      <td>Germline variation at immunoglobulin gene (IG)...</td>\n","      <td>2014-07-03</td>\n","      <td>Corey T Watson; Karyn Meltz Steinberg; Tina A ...</td>\n","      <td>https://doi.org/10.1101/006866</td>\n","      <td>Antibodies are essential components of the imm...</td>\n","      <td>We analyzed sequences from 17 CH17 BAC clones ...</td>\n","    </tr>\n","    <tr>\n","      <th>31178</th>\n","      <td>4da8a87e614373d56070ed272487451266dce919</td>\n","      <td>Bayesian mixture analysis for metagenomic comm...</td>\n","      <td>Deep sequencing of clinical samples is now an ...</td>\n","      <td>2014-07-25</td>\n","      <td>Sofia Morfopoulou; Vincent Plagnol</td>\n","      <td>https://doi.org/10.1101/007476</td>\n","      <td>Metagenomics can be defined as the analysis of...</td>\n","      <td>We first applied metaMix on a popular benchmar...</td>\n","    </tr>\n","    <tr>\n","      <th>31179</th>\n","      <td>eccef80cfbe078235df22398f195d5db462d8000</td>\n","      <td>Mapping a viral phylogeny onto outbreak trees ...</td>\n","      <td>Developing methods to reconstruct transmission...</td>\n","      <td>2014-11-11</td>\n","      <td>Stephen P Velsko; Jonathan E Allen</td>\n","      <td>https://doi.org/10.1101/010389</td>\n","      <td>NaN</td>\n","      <td>: Schematic of transmission links overlaid on ...</td>\n","    </tr>\n","  </tbody>\n","</table>\n","</div>"],"text/plain":["                                            sha  ...                                               body\n","31175  f056da9c64fbf00a4645ae326e8a4339d015d155  ...  Metagenomic classification methods are based o...\n","31176  daf32e013d325a6feb80e83d15aabc64a48fae33  ...  Due to the ubiquity of complex systems in mode...\n","31177  f33c6d94b0efaa198f8f3f20e644625fa3fe10d2  ...  We analyzed sequences from 17 CH17 BAC clones ...\n","31178  4da8a87e614373d56070ed272487451266dce919  ...  We first applied metaMix on a popular benchmar...\n","31179  eccef80cfbe078235df22398f195d5db462d8000  ...  : Schematic of transmission links overlaid on ...\n","\n","[5 rows x 8 columns]"]},"metadata":{"tags":[]},"execution_count":41}]},{"cell_type":"markdown","metadata":{"id":"vIuVL0T7YIv9","colab_type":"text"},"source":["### Download BART pretrained model"]},{"cell_type":"code","metadata":{"id":"KU7HXnU5V0nj","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":17},"outputId":"ef383f8a-065a-4810-d577-8ac0252f81de","executionInfo":{"status":"ok","timestamp":1589121101738,"user_tz":-120,"elapsed":19020,"user":{"displayName":"Bartłomiej Cerek","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GgghwbmOdc_vcrQTrbQaIvP0ASodbJHQUXxGgJyyF4=s64","userId":"03865172822680659937"}}},"source":["tokenizer = BartTokenizer.from_pretrained('bart-large-cnn')\n","model = BartForConditionalGeneration.from_pretrained('bart-large-cnn')\n","model.cuda();"],"execution_count":42,"outputs":[{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}}]},{"cell_type":"code","metadata":{"id":"-_qxxGk9YYI0","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":17},"outputId":"4c4895a9-dd54-47f3-e6c9-59717a01a93d","executionInfo":{"status":"ok","timestamp":1589121101739,"user_tz":-120,"elapsed":19013,"user":{"displayName":"Bartłomiej Cerek","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GgghwbmOdc_vcrQTrbQaIvP0ASodbJHQUXxGgJyyF4=s64","userId":"03865172822680659937"}}},"source":["#Function for abstract genetation\n","def generate_absctract(model, tokenizer, text, max_lenght):\n","    inputs = tokenizer.batch_encode_plus([text], max_length=1024, return_tensors='pt')\n","    summary_ids = model.generate(inputs['input_ids'].cuda(), \n","                                 max_length=max_lenght,\n","                                 early_stopping=True,\n","                                 do_sample=True,\n","                                 top_k=40,\n","                                 top_p=0.95).cpu()\n","    output = [tokenizer.decode(g, skip_special_tokens=True, clean_up_tokenization_spaces=False) for g in summary_ids]\n","    return '\\n'.join(output)"],"execution_count":43,"outputs":[{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}}]},{"cell_type":"markdown","metadata":{"id":"Z0m5uAFg9dAr","colab_type":"text"},"source":["### Generate abstract"]},{"cell_type":"code","metadata":{"id":"IIydPLlH8V4k","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":255},"outputId":"2b5ca28d-f35a-4150-94e1-68d05096fe57","executionInfo":{"status":"ok","timestamp":1589121535731,"user_tz":-120,"elapsed":16505,"user":{"displayName":"Bartłomiej Cerek","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GgghwbmOdc_vcrQTrbQaIvP0ASodbJHQUXxGgJyyF4=s64","userId":"03865172822680659937"}}},"source":["article_no = 3\n","original_text = articles_df.iloc[article_no]['body']\n","original_abstract = articles_df.iloc[article_no]['abstract']\n","original_abstract_len = len(original_abstract.split())\n","\n","print(20*'-', 'ORIGINAL ABSTRACT', 20*'-')\n","print(original_abstract)\n","generatet_abstract = generate_absctract(model, tokenizer, original_text, original_abstract_len)\n","print(20*'-', 'GENERATED ABSTRACT', 20*'-')\n","print(generatet_abstract)"],"execution_count":51,"outputs":[{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"stream","text":["-------------------- ORIGINAL ABSTRACT --------------------\n","Deep sequencing of clinical samples is now an established tool for the detection of infectious pathogens, with direct medical applications. The large amount of data generated provides an opportunity to detect species even at very low levels, provided that computational tools can effectively interpret potentially complex metagenomic mixtures. Data interpretation is complicated by the fact that short sequencing reads can match multiple organisms and by the lack of completeness of existing databases, in particular for viral pathogens. This interpretation problem can be formulated statistically as a mixture model, where the species of origin of each read is missing, but the complete knowledge of all species present in the mixture helps with the individual reads assignment. Several analytical tools have been proposed to approximately solve this computational problem. Here, we show that the use of parallel Monte Carlo Markov chains (MCMC) for the exploration of the species space enables the identification of the set of species most likely to contribute to the mixture. The added accuracy comes at a cost of increased computation time. Our approach is useful for solving complex mixtures involving several related species. We designed our method specifically for the analysis of deep transcriptome sequencing datasets and with a particular focus on viral pathogen detection, but the principles are applicable more generally to all types of metagenomics mixtures. The work is implemented as a user friendly R package, available from CRAN: http://cran.r-project.org/web/packages/metaMix\n","-------------------- GENERATED ABSTRACT --------------------\n","There are three datasets: simHC, simMC, and simMC corresponding to high, medium and low complexity of the metagenomic community respectively. We first applied metaMix on a popular benchmark dataset, for which the community composition and the read assignment is known. We then compared our results with the ones produced by MEGAN version 5.3 and Pathoscope 2.0. We were also interested in comparing our results to the ones by GRAMMy.\n"],"name":"stdout"}]},{"cell_type":"code","metadata":{"id":"gB2oxm34Bl9M","colab_type":"code","colab":{"base_uri":"https://localhost:8080/","height":1000},"outputId":"fd41bd0d-9553-480f-ffff-69eb3f819aa1","executionInfo":{"status":"ok","timestamp":1589121545509,"user_tz":-120,"elapsed":1692,"user":{"displayName":"Bartłomiej Cerek","photoUrl":"https://lh3.googleusercontent.com/a-/AOh14GgghwbmOdc_vcrQTrbQaIvP0ASodbJHQUXxGgJyyF4=s64","userId":"03865172822680659937"}}},"source":["original_text"],"execution_count":52,"outputs":[{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"display_data","data":{"text/html":["\n","  <style>\n","    pre {white-space: pre-wrap;}\n","  </style>\n","  "],"text/plain":["<IPython.core.display.HTML object>"]},"metadata":{"tags":[]}},{"output_type":"execute_result","data":{"text/plain":["'We first applied metaMix on a popular benchmark dataset, for which the community composition and the read assignment is known. We then analyzed RNA-Seq datasets from two clinical samples that were generated for diagnostic purposes. We compare our results with the ones produced by MEGAN version 5.3 and Pathoscope 2.0. Both methods are similarity-based. This property and more specifically their flexibility to work with BLASTx output, makes them better candidates for viral discovery compared to compositionbased methods. From the mixture model methods, we have chosen Pathoscope. We were also interested in comparing our results to the ones by GRAMMy, which was the first similarity-based method to use the idea of the mixture model. However, GRAMMy is designed for nucleotide-nucleotide comparisons (BLASTn), which is suboptimal for viral discovery. GRAMMy also only considers unassembled reads and requires that these are of the same fixed length. For these reasons, GRAMMy was not included in the comparison. Default parameters were used for all methods, unless stated otherwise.\\nFor the metaMix output, we reported organisms with a posterior probability greater than 0.9. The metaMix read support parameter r, which essentially sets the sensitivity/specificity of the method, has an impact on the number of reported species. A large r value can result in the method merging together strains that are differentiated by fewer reads than r. On the other hand a low r can have the opposite effect, whereby the methods splits a strain into two or more strains, by moving a few reads from one strain to a very similar one with which they have equally good matches.\\nThe user\\'s choice for this key parameter r should be informed by the biological context. As an example, for the typical human clinical sample where the sample collection might have occurred some time after the infection has taken place, a low value in order to adopt a sensitive approach is reasonable. Hence, for viral identification in human clinical samples, a low and sensitive value (r = 10) is a reasonable choice. In a highly complex environmental metagenomic community where there is a plethora of species of similar abundances, the choice becomes less straightforward especially in the case of closely related strains. We set the default value for general community profiling in environmental samples to be r = 30. We also compare the output of metaMix for different values of this parameter. Supplementary Text 1 presents a detailed discussion on these settings as well as practical considerations.\\nThe FAMeS artificial datasets (http://fames.jgi-psf.org/description.html), are simulated metagenomic datasets composed of random reads from 113 isolate microbial genomes present in IMG (Integrated Microbial Genomes) and sequenced at the DOE Joint Genome Institute. They are a popular choice to use as benchmark datasets for various metagenomics methods. Their suitability stems from the fact that the number of species that form the metagenomic community is known as well as their relative abundances. The FAMeS datasets have been designed to model real metagenomic communities in terms of complexity and phylogenetic composition.\\nThere are three datasets: simHC, simMC, simLC corresponding to high, medium and low complexity of the metagenomic community respectively. The three methods were applied to simHC, as this is the highest complexity dataset, with many closely related strains with similar abundances and no dominant species. The lowest abundance is 255 reads out of 118,000 reads. The bioinformatics processing in this instance consisted of a BLASTn comparison to all NCBI bacterial genomes (ftp://ftp.ncbi.nlm.nih.gov/ genomes/Bacteria/all.fna.tar.gz). The number of genomes mapped, retrieved from the the BLASTn output was ∼2,500.\\nAs discussed below, metaMix outperforms Pathoscope and MEGAN in the community profiling task (Table 1 ) and consequently in the relative abundance estimation (Table 2) .\\nTo limit the complexity of the fit, we used the two step procedure described in the Methods and fully implemented in metaMix. We first fitted the mixture model with the complete set of 2,500 species and a limited run length of 500 iterations. Based on this analysis, we identified 1,312 species supported by at least one read and explored this state space. To limit the computational time, we also considered a stronger approximation, including only the 374 potential species supported by at least 10 sequencing reads. Both approaches generated similar results, albeit the more complex one with 1,312 potential species required the quadruple of the computation time (12h instead of 3h). metaMix identified 116 species, detecting successfully all the members of the metagenomic community (Supplementary Table S1 ). These were detected on the strain level except in four instances where a different strain of the same species, or different species within the same genus was detected. Four species were identified and not in the simulated dataset, hence can be considered as false positives (Supplementary Table S1 ). enforces the parsimonious results (any thetaPrior greater than 10), thereby removing the unique read penalty, Pathoscope behaves as a standard mixture model and identifies 165 species. With these settings, it identifies all but one members of the community. The organisms are identified at the strain level, except in three instances where it identified different species within the same genus. The major interpretation issue is the presence of a long tail of species (54 species) that are actually not present in the mixture (Supplementary  Table S1 ). Pathoscope produced the results in one minute.\\nMEGAN identified 232 species. It discovered all original species of the community on the strain level, except for 9 instances where it identified the lowest common ancestor (LCA). Aside from the lack of strain or species specificity for 8% of the community members, the main issue is the long tail of false positives species. In the species summary provided by MEGAN, there are 119 species which are not actually present, but supported by a sufficient number of reads (default value: 50 reads) for MEGAN to include these in the output. It finished the computations in less than one minute.\\nThe primary aim for metaMix is to be a diagnostic tool and to answer whether a species is present or absent from the mixture we study. As a secondary aim, we are also interested in estimating accurately the relative abundance of the present organisms. We can assess the abundance estimates produced by the methods by using error measures such as the relative root mean square error, RRMSE and the average relative error, AVGRE. For all methods, when the exact strain was not identified but the correct species or genus was, we used this abundance.\\nwhere t j is the true abundance of species j and w j the estimated abundance. metaMix produces the most accurate abundance estimates and the results are summarized in Table 2 .\\nWe then assessed the importance of the read support parameter r on the output of metaMix. We ran metaMix on the benchmark simHC FAMeS dataset with r = {10, 20, 30, 50} reads (Table 3, Figure 1 ). We observe that as r decreases, a few more related strains from the reference database that are not in the community are retained in the output. As r increases two similar strains are merged into one. We compared these results with the output of Pathoscope and MEGAN. None of these methods have a read support parameter serving the same purpose as in metaMix, so we tuned the most relevant parameters in these tools. Pathoscope has a thetaPrior parameter that enforces a unique read penalty. This parameter represents the read pseudocounts for the non-unique matches and the default setting is zero which allows for non informative priors. Using the default setting Pathoscope identifies 47 taxa. When thetaP\\'s value is in (1,7) it identifies 22 taxa, while with thetaP> 7 it identifies 165. With this latter setting which is the one we chose for the comparison, Pathoscope behaves as a standard mixture model. MEGAN has a \"Min Support\" parameter which sets a threshold for the number of reads that must be assigned to a taxon so that it appears in the result. Any read assigned to a taxon not having the required support is pushed up the taxonomy until a taxon is found that has sufficient support. We used Min support = {10, 20, 30, 50} reads. The respective number of taxa in the summary files were 250, 243, 236, 232.\\nWe then also applied a post-run read count threshold to both methods\\' output summary. We set the threshold for 10,20,30,50 reads respectively, disregarding taxa that have less than that number of reads assigned to them. In all instances metaMix produces a community profile closer to the real one (Table 3, . CC-BY-NC-ND 4.0 International license is made available under a The copyright holder for this preprint (which was not peer-reviewed) is the author/funder. It . https://doi.org/10.1101/007476 doi: bioRxiv preprint Figure 1 ) Pathoscope finds ∼ 20 more false positives while MEGAN ∼ 40 more compared to metaMix at the same read support level.\\nThe simHC from the FAMeS datasets is complex and distinct from typical human clinical samples, putting aside gut microbiome analysis. The differences are the large number of organisms, the presence of closely related strains of similar abundances, as well as the lack of viruses. Nevertheless, it is an essential dataset to use as a benchmark for examining the performance of the methods in a situation of closely related strains in the sample.\\nProtein reference database For the analysis of human clinical samples, we use a custom reference database that combineds viral, bacterial, human and mouse RefSeq proteins. All viruses are used (ftp://ftp.ncbi.nlm.nih.gov/refseq/release/ viral/viral.1.protein.faa.gz) as well as all the bacteria of the human microbiome, according to ftp: //ftp.ncbi.nih.gov/genomes/HUMAN_MICROBIOM/Bacteria/all.faa.tar.gz.\\nTo test metaMix in a clinical setting with a low viral load, we used a brain biopsy RNA-Seq dataset from an undiagnosed encephalitis patient (UCL Hospital, data provided as part of a collaboration with Professor Breuer, UCL). Total RNA was purified from the biopsy and polyA RNA was separated for sequencing library preparation. The Illumina MiSeq instrument generated 20 million paired-end reads. We processed the raw data using the bioinformatics pipeline described in Methods. The processed dataset consisted of ∼ 75, 000 non-host reads and contigs. Based on the BLASTx output there were 1,298 potential species.\\nFollowing the initial processing, we used metaMix for species identification and abundance estimation. The resulting species profile is shown in Table 4 ; the 13 metaMix entries correspond to 10 species. The most abundant organism was the φX174 bacteriophage, which is routinely used for deep-sequencing quality control. More interestingly, we identified an astrovirus. Five short assembled contigs (44 reads) with length ranging between 167bp and 471bp and two non-assembled reads were assigned to the Astrovirus VA1 with a probability score of 1 (Figure 2) . metaMix also identified a number of bacteria supported by a few reads. These are either known laboratory reagent contaminants or human skin associated contaminants (Salter et al., 2014) . The analysis completed in 29 minutes.\\nThe presence of the astrovirus was confirmed using real-time RT-PCR. Genome sequencing of the astrovirus in the sample and subsequent study of the consensus sequence showed that we had in fact identified a novel virus, closely related to the VA1 strain ( (Brown et al., 2014) , in press).\\nPathoscope identified 22 taxa, corresponding to 15 species and some genera or families (Table 4) . It also assigned all 46 reads to the Astrovirus VA1. Almost all the species identified from metaMix were identified by Pathoscope, with an additional 9 taxa supported by few reads. As the method can only properly work with unassembled sequence data, an extra BLASTx similarity step had to be performed for the 91,516 reads that had contributed to the 679 assembled contigs. Pathoscope produced the results in less than one minute.\\nMEGAN identified 19 taxa and did not detect the astrovirus signal. We modified the minimum read support parameter from 50 reads to 10 to increase sensitivity. MEGAN then identified 25 taxa, including the Astrovirus VA1. The remaining 24 were mostly genera, relevant to the species detected by metaMix and Pathoscope. MEGAN produced the results in less than one minute. The copyright holder for this preprint (which was not peer-reviewed) is the author/funder. It . https://doi.org/10.1101/007476 doi: bioRxiv preprint\\nHuman clinical sample -species absent from the database We then compared the performance of the three methods in a scenario where sequences present in the sample are absent from our reference database. We analyzed a second brain biopsy sample from an undiagnosed patient. 32 million RNA-Seq reads were obtained using the HiSeq instrument. Following initial processing using our bionformatics pipeline, the dataset had 1,261,575 non host reads and contigs for subsequent analyses. There were 3,150 potential species based on the BLASTx output.\\nThe resulting species profile consisted of 7 species (Supplementary Table S2 ). The most interesting finding was the identification of a human coronavirus. We found two different strains: Human coronavirus OC43 had almost a million reads assigned to it. Additionally there were 67K reads assigned to Human enteric coronavirus strain 4408. The presence of both viral strains in the results indicated that even though the virus in the sample was mostly similar to the OC43 strain, there were sequences sharing higher similarity to 4408 at some loci. This is highlighting how the database choice impacts the results: the RefSeq database we used has only one OC43 strain, while in GenBank there are several OC43 strains capturing the high mutation rates of the Human Coronavirus OC43.\\nWe followed up on the sequences assigned to the \"unknown category\", that is approximately 170K reads, looking for nucleotide similarity with NR-NT using BLASTn. Approximately half of the reads originated from an untranslated region of the Coronavirus genome, which is not captured by the protein reference database. The remaining reads matched confidently to either Danio rerio (zebrafish) sequences or Gallus gallus (chicken), two organisms whose proteins are not in the human microbiome reference we are using. The zebrafish and chicken matches were explained as barcode leakage resulting from multiplexing on the same flowcell zebrafish and chicken RNA-Seq libraries. metaMix appropriately assigned these reads to the \"unknown\" category, producing a clean probabilistic summary (Supplementary Table S2 ). The method ran in 4.7 hours.\\nIn this instance, the metaMix results emphasize the importance of being able to deal with missing reference sequences that do not have a closely related strain or species in the same database.\\nPathoscope identified 177 species in this sample (Supplementary Table S2 ). We optimized the value of the unique read penalty parameter and we achieved the best results with the thetaPrior parameter set within the range 10-100. With these settings, the method identified 52 species (Supplementary Table S2 ). Our assessment is that Pathoscope is confused by the lack of completeness of databases combined with the absence of an \"unknown\" category, which prevents it from dealing with these unassigned reads sensibly. Pathoscope completed its analysis in 10 minutes.\\nMEGAN assigned the reads to 30 taxa. These included some species and genera but most were families (Supplementary Table S2 ). Approximately 250K reads could not be assigned to any taxonomic level. MEGAN run in 8 minutes.\\nHere, we present metaMix, a sensitive method for metagenomic species identification and abundance estimation. The method is implemented in an R package (http://cran.r-project.org/web/packages/metaMix). Using a Bayesian mixture model framework, we account for model uncertainty by performing model averaging and we resolve ambiguous assignments by considering all reads simultaneously. A key feature of the method is that it provides probabilities that answer pertinent biological questions, in particular the posterior probability for the presence of a species in the mixture. Additionally it accurately quantifies the relative proportions of the organisms. The copyright holder for this preprint (which was not peer-reviewed) is the author/funder. It . https://doi.org/10.1101/007476 doi: bioRxiv preprint This general framework is designed to address interpretation issues associated with closely related strains in the sample, low abundance organisms and absence of genomes from the reference database. We show that metaMix outperforms other methods in the community profiling task, particularly when complex structures with closely related strains are studied. As a consequence, it also produces more accurate relative abundance estimates for the species in the mixture. The method can deal with either unassembled reads or assembled contigs or both, allowing for flexibility of choice for the bioinformatics preprocessing. In practice, the choice of bioinformatics processing prior to the application of our Bayesian mixture analysis must be optimized for each application, and our processing pipeline has been designed with viral sequence identification from transcriptome sequencing as a main goal. Nevertheless, as demonstrated by our analysis of the mock bacterial community dataset, the method can be applied in other contexts.\\nThe sensitivity and general applicability of metaMix comes at an increased computational cost, requiring access to a multi-core computer to run efficiently. For the datasets presented here, the computation time remained manageable and did not exceed a few hours, using 12 cores to run 12 parallel chains. Nevertheless, a limitation of metaMix is the increased processing time for very large datasets. Speed related improvements can be implemented in scenarios where the species ambiguity concerns only a small proportion of the read set. Reads with certain assignments can be flagged prior to the MCMC exploration of the state-space. Their assignment information can then be carried forward, thereby reducing the size of the similarity matrix used as input by the mixture model. Another area of possible improvement is MCMC convergence determination. The current version of metaMix produces log-likelihood traceplots allowing the user to visually inspect the MCMC convergence, however additional diagnostic criteria can be implemented in future versions. metaMix is most useful for complex datasets for which the interpretation is challenging. It has been mainly used as a clinical diagnostic tool, helping with the identification of the infecting pathogen while providing an accurate profile of the community in the sample.\\nPrior to running the mixture model for metagenomic profiling, several steps are required to process the short read sequence data (Figure 3 ). The pipeline uses publicly available bioinformatics tools for each preprocessing step.\\nThe first step is the removal of clonal reads using an in house C++ script. We then use PRINSEQ (Schmieder and Edwards, 2011) for read-based quality control, removing low quality and complexity reads and performing 3\\'end trimming. For metagenomic analysis of human samples, reads originating from the human host are not relevant for our research question. We therefore remove human host reads, using a twostep approach to limit computation time: initially a short read aligner (novoalign, www.novocraft.com), followed by BLASTn. The next step is only applicable when the focus is on virus discovery using transcriptome reads. We remove ribosomal RNA sequences, using BLASTn against the Silva rRNA database (http://www.arb-silva.de/).\\nThe remaining reads are assembled into contigs using the Velvet short read assembler (Zerbino and Birney, 2008) . For each contig we record the number of reads required for its assembly, using this information at the stage of species abundance estimation. A Velvet tuning parameter is the user defined k-mer length that specifies the extent of overlap required to assemble read pairs. Metagenomic assembly is not a straightforward task, as short k-mers work best with the low abundance organisms, while long k-mers with the highly abundance ones. The shorter the k-mer the greater the chance of spurious overlaps, hence we choose relatively high k-mer length, in order to avoid chimeric contigs.\\nFor each contig and unassembled read we record the potentially originating species, using the nucleotide to protein homology matching tool BLASTx. We use BLASTx due to the higher level of conservation expected at the protein level compared to nucleotides. This choice is guided by our focus in viral pathogens -viruses having high genetic diversity and divergence (Fancello et al., 2012) .\\nThis step generates a sparse similarity matrix between the read sequences and the protein sequences, with species as columns, reads and contigs as rows.\\nThe statistical method described in the remainder of this section considers the competing models that could accommodate our observed data, that is the BLASTx results and compares them. The different models . CC-BY-NC-ND 4.0 International license is made available under a The copyright holder for this preprint (which was not peer-reviewed) is the author/funder. It . https://doi.org/10.1101/007476 doi: bioRxiv preprint represent different sets of species being present in the sample. The method works on two levels of inference: in the first instance we assume a set of species to be present in the sample and we estimate this model\\'s parameters given the data. The other level of inference is the model comparison so as to assess the more plausible model. The process is iterated in order to explore the model state space.\\nAssuming a given set of K species from which the reads can originate, the metagenomic problem can be summarized as a mixture problem, for which the assignment of the sequencing reads to species is unknown and must be determined. The data consist of N sequencing reads X = (x 1 , . . . , x N ), and for a given read x i the likelihood is written as:\\nwhere w = (w 1 , ..., w K ) represent the proportion of each of the K species in the mixture. These mixture weights are constrained such that 0 ≤ w j ≤ 1 and j w j = 1. In practice, we also add a category (species K + 1) which we refer to as the \"unknown\" category, and captures the fact that some reads cannot be assigned to any species.\\nAdditionally f j (x i ) = P (x i |x i from species j) = p ij is the probability of observing the read x i conditional on the assumption that it originated from species j. We model this probability using the number of mismatches m between the translated read sequence and the reference sequence and a Poisson distribution with parameter λ for that number of mismatches.\\nwhere l g is the length of the reference genome, when short reads are matched to a nucleotide database. For nucleotide matching, l g has a large impact on the probability computation. However, when matching against protein databases, the more limited heterogeneity of protein lengths results in a much smaller impact of the length parameter. In addition, incomplete annotation can potentially make the inclusion of protein length problematic for the p ij computation. Consequently, for protein matched sequences, we simply defined our p ij as:\\nTherefore for a given set of K species, the p ij probabilities are regarded as known and the mixture weights must be estimated. Combining the above we see that when we know the set of species K, the mixture distribution gives the probability of observing read x i : K j=1 w j p ij , namely equation (3). We therefore write the likelihood of the dataset X as a sum of K n terms:\\nAssuming a fixed set of species, the posterior probability distribution of the weights w given the read data X is:\\nA practical prior for the mixing parameters w is the Dirichlet distribution owing to its conjugate status to the multinomial distribution. Despite the use of conjugate priors, the probabilistic assignment of reads to species involves the expansion of the likelihood into K n terms which is computationally infeasible through direct computation. An efficient estimation can be performed by the introduction of unobserved latent variables that code for the read assignments. In this framework, either the Gibbs sampler (Marin et al., 2005) , a Monte Carlo Markov Chain technique, or the Expectation-Maximization (EM) (Dempster and . CC-BY-NC-ND 4.0 International license is made available under a The copyright holder for this preprint (which was not peer-reviewed) is the author/funder. It . https://doi.org/10.1101/007476 doi: bioRxiv preprint Laird, 1977) algorithm can be used to estimate the mixture weights w. EM returns a point estimate for w while the Gibbs sampler the distribution of w (Supplementary Text 1 for details of implementation). Both methods were implemented and provided comparable results.\\nEach combination of species corresponds to a finite mixture model for which the marginal likelihood can be estimated. Marginal likelihood comparison has a central role in comparing different models {M 1 , . . . , M m }. To compute the marginal likelihood P (X|M k ) for the mixture model M k one has to average over the parameters with respect to the prior distribution π(θ k |M k ), where θ k are the model parameters:\\nThe posterior probability of the model M k is:\\nwhere P (M k ) is the prior belief we hold for each model. The prior can be specified depending on the context but the basis of our interpretation is that parsimonious models with a limited number of species are more likely. Thus in this Bayesian framework our default prior uses a penalty limiting the number of species in the model.\\nWe approximate this penalty factor based on a user-defined parameter r that represents the species read support required by the user to believe in the presence of this species. We compute the logarithmic penalty value as the log-likelihood difference between two models: one where all N reads belong to the \"unknown\" category and one where r reads have a perfect match to some unspecified species and the remaining N − r reads belong to the \"unknown\" category. In the nucleotide similarity situation, the p ij probabilities for the r reads originating from this unspecified species are approximated by 1/(median genome length in the reference database). This parameter essentially reflects how many reads are required to provide credible support that a species is present in the mixture and acts as a probabilistic threshold as opposed to a deterministic one applied on a ranked list.\\nFrom now on, when we refer to the marginal likelihood, we mean the marginal likelihood for a specific model and we forego conditioning on the model M k in the notation. Additionally, in our mixture model p ij are always regarded as known, therefore the model parameters θ k are the mixture weights w. Hence (8) becomes:\\nP (X) = w P (X|w)π(w)dw\\nApproximating the marginal likelihood is a task both difficult and time-consuming. We chose the Defensive Importance Sampling technique (Hesterberg, 1995) for the relative simple implementation compared to other approaches (Supplementary Text 1 for details of implementation). This is crucial as we perform this approximation numerous times, for every species combination we consider.\\nHowever the goal of this work is to deliver results in a clinical setting within an actionable time-frame. We wish to speed up the computation without compromising the accuracy and the sensitivity of the results. For that reason, we use a point estimate of the marginal likelihood by means of the Expectation-Maximization (EM) algorithm. The different approaches were used on the benchmark dataset. The resulting taxonomic assignment as well as the species relative abundance estimates were similar between them, with the EM approach resulting in a 13-fold speed increase (Supplementary Text 1).\\nWe use a Monte Carlo Markov Chain (MCMC) to explore the set of present species of size 2 S − 1, where S is the total number of potential species. In practice we observe that S can be greater than 1,000. The MCMC must explore the state-space in a clinically useful timespan. Therefore we reduce the size of the state-space, by decreasing the number of S species to the low hundreds. We achieve this by fitting a mixture model . CC-BY-NC-ND 4.0 International license is made available under a The copyright holder for this preprint (which was not peer-reviewed) is the author/funder. It . https://doi.org/10.1101/007476 doi: bioRxiv preprint with S categories, considering all potential species simultaneously. Post fitting, we retain only the species categories that are not empty, that is categories that have at least one read assigned to them.\\nLet us assume that at step t, we deal with a set of species that corresponds to the mixture model M k . At the next step (t + 1), we either add or remove a species and the new set corresponds to the mixture model M l . The step proposing the model M l is accepted with probability:\\nwhere q(M l → M k ) is the probability of transitioning from model M l to model M k . In other words, this is the probability of adding or removing the species to the M k set of species that took us to the M l set of species.\\nIf the step is accepted, then the chain moves to the new proposed state M l . Otherwise if not accepted, the chain\\'s current state becomes the previous state of the chain, i.e the set of species remains unchanged. metaMix outputs log-likelihood traceplots so that the user can visually inspect the mixing and the convergence of the chain. The default setting is to discard the first 20% of the iterations as burn-in. We concentrate on the rest to study the distribution over the model choices and perform model averaging. We can then summarize appropriately the posterior distribution and answer the important questions of interest. Examples of such questions include: what species have probability p or greater being included in the set of present species? what is the probability of having the n specific closely related strains in the set of present species? Depending on the biological context, one may ask numerous similar or other case-specific questions.\\nWe observed that simple MCMC does not efficiently explore the complex model state space, as evidenced by the poor mixing of the chain (Figure 4) .\\nIn order to overcome this and take advantage of parallel computing, we run multiple chains and allow exchange moves between them. This method is called parallel tempering MCMC (Earl and Deem, 2005) . Within the parallel setting, each chain simulates from the posterior distribution g(M )=P (M |X) raised to a temperature t (0, 1], where model M M represents a set of species being present. The different temperature levels result in tempered versions of the posterior distribution P (M k |X) t=1/T . When T = 1 the draws are from the posterior distribution. On the other hand, at higher temperatures the posterior spreads out its mass and becomes flatter. In practice that means that distributions at higher temperatures are easily sampled, improving the mixing. We are interesting in studying the original posterior distribution with T = 1.\\nWe implemented two types of moves. The first is the mutation step, which simply is the within chain move we described in the previous section. This is accepted with probability given by (11). The other is the exchange step, a between chains move. This Metropolis-Hastings move proposes to swap the value of two chains k and k + 1, adjacent in terms of T . Suppose that the values of the two chains are M k and M k+1 respectively, corresponding to two different sets of species. The move is accepted with probability (Jasra et al., 2007) :\\nSince g k (M k )=P (M k |X) T and g k+1 (M k+1 ) = P (M k+1 |X) T +1 , it follows that when M k+1 represents a set of species of higher probability than the one M k represents, the exchange move will always be accepted (Figure 4) .\\nThis allows moves between separate modes, ensuring a global exploration of the model state space. Eventually \"hot\" and \"cold\" chains will progress towards a global mode.\\nthe Annals of Human Genetics. This work is also supported by the National Institute for Health Research University College London Hospitals Biomedical Research Centre. For metaMix that is r={10, 20, 30, 50} reads, for Pathoscope thetaPrior> 7+ post-run threshold ={10, 20, 30, 50} reads, for MEGAN \"Min Support\" + post-run threshold ={10, 20, 30, 50} reads. The copyright holder for this preprint (which was not peer-reviewed) is the author/funder. The copyright holder for this preprint (which was not peer-reviewed) is the author/funder. It . https://doi.org/10.1101/007476 doi: bioRxiv preprint Tables . CC-BY-NC-ND 4.0 International license is made available under a The copyright holder for this preprint (which was not peer-reviewed) is the author/funder. It . https://doi.org/10.1101/007476 doi: bioRxiv preprint Staphylococcus epidermidis --10 28211\\nAlphaproteobacteria --10 28037\\nStreptococcus mitis --8 562\\nEscherichia coli --8 509173\\nAcinetobacter baumannii AYE --7 41297\\nSphingomonadaceae --6 40214\\nAcinetobacter johnsonii --6 29391\\nGemella morbillorum --5 76122\\nAlloprevotella tannerae --4 652103\\nRhodopseudomonas palustris DX-1 --2 268747\\nProchlorococcus phage P-SSM4 --2 . CC-BY-NC-ND 4.0 International license is made available under a The copyright holder for this preprint (which was not peer-reviewed) is the author/funder. It . https://doi.org/10.1101/007476 doi: bioRxiv preprint'"]},"metadata":{"tags":[]},"execution_count":52}]}]}